### PR TITLE
Auth: Support google OIDC and group fetching

### LINKED ISSUE #69520: Auth: Google OAuth default configuration are different from .well-known/openid-configuration
**What is the chore?**:

- Google OAuth default configuration are different from .[well-known/openid-configuration](https://accounts.google.com/.well-known/openid-configuration)
  - For now the defaults.ini URL will work, but better choose the same endpoints that google provides in compliance with OIDC standards.
  - If implement OIDC in the future, it will only a small change

|| defaults.ini           | /.well-known/openid-configuration             |
|------------------------|-----------------------------------------------|--------------------------------------------------|
| token_endpoint         | https://accounts.google.com/o/oauth2/token    | https://oauth2.googleapis.com/token              |
| authorization_endpoint | https://accounts.google.com/o/oauth2/auth     | https://accounts.google.com/o/oauth2/v2/auth     |
| userinfo_endpoint      | https://www.googleapis.com/oauth2/v1/userinfo | https://openidconnect.googleapis.com/v1/userinfo |
| scope | https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email | openid email profile |

**Is there anything else we need to know?**:

see: https://accounts.google.com/.well-known/openid-configuration

- suggest labels
  - https://github.com/grafana/grafana/labels/area%2Fauth%2Foauth
  - https://github.com/grafana/grafana/labels/team%2Fgrafana-authnz

```json
{
 "issuer": "https://accounts.google.com",
 "authorization_endpoint": "https://accounts.google.com/o/oauth2/v2/auth",
 "device_authorization_endpoint": "https://oauth2.googleapis.com/device/code",
 "token_endpoint": "https://oauth2.googleapis.com/token",
 "userinfo_endpoint": "https://openidconnect.googleapis.com/v1/userinfo",
 "revocation_endpoint": "https://oauth2.googleapis.com/revoke",
 "jwks_uri": "https://www.googleapis.com/oauth2/v3/certs",
 "response_types_supported": [
  "code",
  "token",
  "id_token",
  "code token",
  "code id_token",
  "token id_token",
  "code token id_token",
  "none"
 ],
 "subject_types_supported": [
  "public"
 ],
 "id_token_signing_alg_values_supported": [
  "RS256"
 ],
 "scopes_supported": [
  "openid",
  "email",
  "profile"
 ],
 "token_endpoint_auth_methods_supported": [
  "client_secret_post",
  "client_secret_basic"
 ],
 "claims_supported": [
  "aud",
  "email",
  "email_verified",
  "exp",
  "family_name",
  "given_name",
  "iat",
  "iss",
  "locale",
  "name",
  "picture",
  "sub"
 ],
 "code_challenge_methods_supported": [
  "plain",
  "S256"
 ],
 "grant_types_supported": [
  "authorization_code",
  "refresh_token",
  "urn:ietf:params:oauth:grant-type:device_code",
  "urn:ietf:params:oauth:grant-type:jwt-bearer"
 ]
}
```

### PR DESCRIPTION
**Pull request by Jguer**
**What is this feature?**

- Add group fetching and teamsync support to `auth.google`
- Support google OIDC endpoints 

Closes https://github.com/grafana/grafana/issues/69520
Closes https://github.com/grafana/grafana/pull/69521
Implements https://github.com/grafana/grafana/issues/70081

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/login/social/google_oauth.go [hunk 1]
```diff
@@ -48,3 +116,95 @@ func (s *SocialGoogle) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption)
...
+	}
+
+	s.log.Debug("Retrieving groups", "scopes", s.SocialBase.Config.Scopes)
+	if !contains(s.Scopes, "https://www.googleapis.com/auth/cloud-identity.groups.readonly") {
+		return nil, nil
+	}
+
+	groups := []string{}
+
+	url := fmt.Sprintf("%s?query=member_key_id=='%s'", googleIAMGroupsEndpoint, userData.Email)
```
Comment (jmatosgrafana) [inline pkg/login/social/google_oauth.go:173]: Because Email has been provided by Google I don't believe it could be manipulated to inject unexpected characters in the url

**Review by jmatosgrafana — APPROVED** [review]
LGTM from a security point of view

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 2]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
 # ..
 skip_org_role_sync = true
 ```
+
+### Configure team sync for Google OAuth
+
+> Available in Grafana v10.1.0 and later versions.
+
+Team sync allows you to automatically add users to teams based on their Google groups. The following example shows how to configure team sync for Google OAuth.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:107]:
```suggestion
With team sync, you can easily add users to teams by utilizing their Google groups. To set up team sync for Google OAuth, refer to the following example.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 3]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
...
+
+1. Enable Cloud Identity API on your [organization's dashboard](https://console.cloud.google.com/apis/api/cloudidentity.googleapis.com/)
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:109]:
```suggestion
1. Enable the Google Cloud Identity API on your [organization's dashboard](https://console.cloud.google.com/apis/api/cloudidentity.googleapis.com/).
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 4]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
...
+
+2. Add the `https://www.googleapis.com/auth/cloud-identity.groups.readonly` scope to your Grafana `[auth.google]` configuration:
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:111]:
```suggestion
1. Add the `https://www.googleapis.com/auth/cloud-identity.groups.readonly` scope to your Grafana `[auth.google]` configuration:
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 5]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
...
+
+Example:
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:113]: indent

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 6]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
...
+
+```ini
+[auth.google]
+# ..
+scopes = openid email profile https://www.googleapis.com/auth/cloud-identity.groups.readonly
+```
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:119]: indent

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 7]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
...
+
+3. Configure team sync in your Grafana team's `External group sync` tab.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:121]:
```suggestion
1. Configure team sync in your Grafana team's `External group sync` tab.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 8]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
...
+   The external group ID for a Google group is the group's email address, such as `dev@grafana.com`
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:122]:
```suggestion
   The external group ID for a Google group is the group's email address, such as `dev@grafana.com`.
```

**Review by chri2547 — CHANGES_REQUESTED** [review]
A few suggestions. Thank you for the contribution!

File: pkg/login/social/google_oauth.go [hunk 9]
```diff
@@ -48,3 +116,95 @@ func (s *SocialGoogle) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption)
...
+		return nil, nil
+	}
+
+	rawJSON, err := s.retrieveRawIDToken(idToken)
+	if err != nil {
+		s.log.Warn("Error retrieving id_token", "error", err, "token", fmt.Sprintf("%+v", idToken))
+		return nil, nil
+	}
+
+	s.log.Debug("Received id_token", "raw_json", string(rawJSON))
```
Comment (mgyongyosi) [inline pkg/login/social/google_oauth.go:135]: I think it should be logged in DevMode only (as with the access_token) and/or remove the values from the json or anonymize those. 

What do you think? I think the id_token could contain PII information.

File: pkg/login/social/google_oauth.go [hunk 10]
```diff
@@ -48,3 +116,95 @@ func (s *SocialGoogle) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption)
...
+		GroupKey struct {
+			ID string `json:"id"`
+		} `json:"groupKey"`
+		DisplayName string `json:"displayName"`
+	} `json:"memberships"`
+	NextPageToken string `json:"nextPageToken"`
+}
+
+func (s *SocialGoogle) retrieveGroups(ctx context.Context, client *http.Client, userData *googleUserData) ([]string, error) {
+	contains := func(slice []string, item string) bool {
```
Comment (mgyongyosi) [inline pkg/login/social/google_oauth.go:157]: Nit: [slices.Contains(..)](https://pkg.go.dev/golang.org/x/exp/slices#Contains)

File: pkg/login/social/google_oauth.go [hunk 11]
```diff
@@ -11,35 +11,103 @@ import (
...
+type googleAPIData struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"verified_email"`
+}
+
+func (s *SocialGoogle) extractFromAPI(ctx context.Context, client *http.Client) (*googleUserData, error) {
+	if legacyAPIURL == s.apiUrl {
+		s.log.Warn("Using legacy Google API URL, please update your configuration")
```
Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth.go:81]: nit: should we log this somewhere else? Maybe where the config is parsed. Otherwise this log will be pretty spammy.

File: pkg/login/social/google_oauth.go [hunk 12]
```diff
@@ -11,35 +11,103 @@ import (
...
+		data := googleAPIData{}
+		response, err := s.httpGet(ctx, client, s.apiUrl)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling legacy user info: %s", err)
```
Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth.go:85]:
```suggestion
			return nil, fmt.Errorf("error retrieving legacy user info: %s", err)
```

File: pkg/login/social/google_oauth.go [hunk 13]
```diff
@@ -11,35 +11,103 @@ import (
...
+
+type googleAPIData struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"verified_email"`
+}
+
+func (s *SocialGoogle) extractFromAPI(ctx context.Context, client *http.Client) (*googleUserData, error) {
+	if legacyAPIURL == s.apiUrl {
```
Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth.go:80]: thought: should we check that legacyAPIURL is the prefix of `s.apiUrl`? In case the user has specified `https://www.googleapis.com/oauth2/v1/userinfo/` as the URL.

Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth.go:80]:
```suggestion
	if strings.HasPrefix(s.apiUrl, legacyAPIURL) {
```

File: pkg/login/social/google_oauth.go [hunk 14]
```diff
@@ -48,3 +116,95 @@ func (s *SocialGoogle) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption)
...
+	rawJSON, err := s.retrieveRawIDToken(idToken)
+	if err != nil {
+		s.log.Warn("Error retrieving id_token", "error", err, "token", fmt.Sprintf("%+v", idToken))
+		return nil, nil
+	}
+
+	s.log.Debug("Received id_token", "raw_json", string(rawJSON))
+
+	var data googleUserData
+	if err := json.Unmarshal(rawJSON, &data); err != nil {
```
Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth.go:138]: question: is it possible that the ID token is in the legacy format?

File: pkg/login/social/google_oauth.go [hunk 15]
```diff
@@ -48,3 +116,95 @@ func (s *SocialGoogle) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption)
...
+		for _, s := range slice {
+			if s == item {
+				return true
+			}
+		}
+		return false
+	}
+
+	s.log.Debug("Retrieving groups", "scopes", s.SocialBase.Config.Scopes)
+	if !contains(s.Scopes, "https://www.googleapis.com/auth/cloud-identity.groups.readonly") {
```
Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth.go:167]: nit: would `https://www.googleapis.com/auth/admin.directory.group` not be a valid scope as well? It's excessive, but should still work, right?

Comment (IevaVasiljeva) [reply pkg/login/social/google_oauth.go:167]: Also should we add a debug log for this case?

File: pkg/login/social/google_oauth_test.go [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+				},
+				userData: &googleUserData{
+					Email: "test@example.com",
+				},
+			},
+			want:    []string{"test-group@google.com", "test-group-2@google.com"},
+			wantErr: false,
+		},
+		{
+			name: "Groups",
```
Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth_test.go:135]:
```suggestion
			name: "Has one page to fetch",
```

File: pkg/login/social/google_oauth_test.go [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+				client: &http.Client{},
+				userData: &googleUserData{
+					Email: "test@example.com",
+				},
+			},
+			want:    nil,
+			wantErr: false,
+		},
+		{
+			name: "No groups",
```
Comment (IevaVasiljeva) [inline pkg/login/social/google_oauth_test.go:50]: This test is currently the same as above - should add a scope and return empty list of groups.

**Review by IevaVasiljeva — COMMENTED** [review]
Mostly looks good, but I've left some comments for edge cases that might need to be covered.

File: pkg/login/social/google_oauth.go (hunk 10 above)
Comment (Jguer) [reply pkg/login/social/google_oauth.go:157]: :smile: just made me feel really old right now

File: pkg/login/social/google_oauth.go (hunk 11 above)
Comment (Jguer) [reply pkg/login/social/google_oauth.go:81]: Agree

File: pkg/login/social/google_oauth.go (hunk 14 above)
Comment (Jguer) [reply pkg/login/social/google_oauth.go:138]: seems not, it's a openID standard compliant https://openid.net/specs/openid-connect-core-1_0.html#IDToken

File: pkg/login/social/google_oauth.go (hunk 15 above)
Comment (Jguer) [reply pkg/login/social/google_oauth.go:167]: I will try it out, I think it doesn't as it's meant for a different API from my last test but let me try again

File: pkg/login/social/google_oauth_test.go (hunk 17 above)
Comment (Jguer) [reply pkg/login/social/google_oauth_test.go:50]: my bad, must have wrapped up for the day and forgot it

File: pkg/login/social/google_oauth.go (hunk 15 above)
Comment (Jguer) [reply pkg/login/social/google_oauth.go:167]:
```
DEBUG[06-16|15:16:31] Retrieving groups                        logger=oauth.google scopes="[openid profile email https://www.googleapis.com/auth/admin.directory.group]"
WARN [06-16|15:16:31] Error retrieving groups                  logger=oauth.google error="error getting groups: unsuccessful response status code 403: {\n  \"error\": {\n    \"code\": 403,\n    \"message\": \"Request had insufficient authentication scopes.\",\n    \"status\": \"PERMISSION_DENIED\",\n    \"details\": [\n      {\n        \"@type\": \"type.googleapis.com/google.rpc.ErrorInfo\",\n        \"reason\": \"ACCESS_TOKEN_SCOPE_INSUFFICIENT\",\n        \"domain\": \"googleapis.com\",\n        \"metadata\": {\n          \"method\": \"google.apps.cloudidentity.groups.v1.MembershipsService.SearchDirectGroups\",\n          \"service\": \"cloudidentity.googleapis.com\"\n        }\n      }\n    ]\n  }\n}\n"
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 18]
```diff
@@ -98,3 +99,26 @@ We do not currently sync roles from Google and instead set the AutoAssigned role
 # ..
 skip_org_role_sync = true
 ```
+
+### Configure team sync for Google OAuth
+
+> Available in Grafana v10.1.0 and later versions.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:105]: Because we version our docs, we don't need to add this kind of note. You can remove it.

Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:105]:
```suggestion
```

**Review by chri2547 — APPROVED** [review]
One additional suggestion and approved. Thank you for the contribution!

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md (hunk 18 above)
Comment (Jguer) [reply docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:105]: Hey @chri2547 I'm ok with removing but we end up getting PRs to re-add as users are used to using the latest version of the documentation. Could you provide further guidance?

Example: https://github.com/grafana/grafana/pull/69419

Comment (chri2547) [reply docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:105]: Hi @Jguer , I checked with the team because you made a good point that I hadn't considered. Our guidance is to keep these kinds of notes in, and when they get stale (for example, 9.5 docs refer to 7.x), remove them as we come across them.

**Review by chri2547 — APPROVED** [review]

**Review by mgyongyosi — APPROVED** [review]
